Sources/Sentry/SentryThreadInspector.m

```diff
@@ -28,27 +64,89 @@ - (id)initWithStacktraceBuilder:(SentryStacktraceBuilder *)stacktraceBuilder
 {
     NSMutableArray<SentryThread *> *threads = [NSMutableArray new];
 
-    SentryCrashMC_NEW_CONTEXT(context);
-    [self.machineContextWrapper fillContextForCurrentThread:context];
+    @synchronized(self) {
+        SentryCrashMC_NEW_CONTEXT(context);
+        SentryCrashThread currentThread = sentrycrashthread_self();
 
-    int threadCount = [self.machineContextWrapper getThreadCount:context];
+        [self.machineContextWrapper fillContextForCurrentThread:context];
+        int threadCount = [self.machineContextWrapper getThreadCount:context];
 
-    for (int i = 0; i < threadCount; i++) {
-        SentryCrashThread thread = [self.machineContextWrapper getThread:context withIndex:i];
-        SentryThread *sentryThread = [[SentryThread alloc] initWithThreadId:@(i)];
+        for (int i = 0; i < threadCount; i++) {
+            SentryCrashThread thread = [self.machineContextWrapper getThread:context withIndex:i];
+            SentryThread *sentryThread = [[SentryThread alloc] initWithThreadId:@(i)];
 
-        sentryThread.name = [self getThreadName:thread];
+            sentryThread.name = [self getThreadName:thread];
 
-        sentryThread.crashed = @NO;
-        bool isCurrent = thread == sentrycrashthread_self();
-        sentryThread.current = @(isCurrent);
+            sentryThread.crashed = @NO;
+            bool isCurrent = thread == currentThread;
+            sentryThread.current = @(isCurrent);
 
-        // For now we can only retrieve the stack trace of the current thread.
-        if (isCurrent) {
-            sentryThread.stacktrace = [self.stacktraceBuilder buildStacktraceForCurrentThread];
+            if (isCurrent) {
+                sentryThread.stacktrace = [self.stacktraceBuilder buildStacktraceForCurrentThread];
+            }
+
+            // We need to make sure the main thread is always the first thread in the result
+            if ([self.machineContextWrapper isMainThread:thread])
+                [threads insertObject:sentryThread atIndex:0];
+            else
+                [threads addObject:sentryThread];
         }
+    }
 
-        [threads addObject:sentryThread];
+    return threads;
+}
+
+- (NSArray<SentryThread *> *)getCurrentThreadsWithStackTrace
+{
+    NSMutableArray<SentryThread *> *threads = [NSMutableArray new];
+
+    @synchronized(self) {
+        SentryCrashMC_NEW_CONTEXT(context);
+        SentryCrashThread currentThread = sentrycrashthread_self();
+
+        thread_act_array_t suspendedThreads = NULL;
+        mach_msg_type_number_t numSuspendedThreads = 0;
+
+        sentrycrashmc_suspendEnvironment(&suspendedThreads, &numSuspendedThreads);
+
+        ThreadInfo threadsInfos[numSuspendedThreads];
+
+        for (int i = 0; i < numSuspendedThreads; i++) {
+            if (suspendedThreads[i] != currentThread) {
```

Comment:
`m`: Why can't we do this if it's the `currentThread` and why do we have to get the stacktrace with `[self.stacktraceBuilder buildStacktraceForCurrentThread]` below?

Reply:
Maybe we can. I didn't want to touch the code that much since I know for the current thread this works.